### PR TITLE
Add content-based file type detection

### PR DESCRIPTION
**Pull request by alumi**
#### summary

`cljam.io.sam/reader`, `cljam.io.vcf/reader` and `cljam.io.sequence/reader` determines a file type based on a suffix of the input file. This can be problematic when the suffix is not available. (e.g. hashed filenames for temporary use)
This PR adds fallbacks for detecting file types.
When failed to detect a file type from the suffix, those functions now try to detect it based on the content of the file.

#### tests
- `lein check` 🆗 
- `lein test :all` 🆗 
- `lein cloverage` 🆗 
- `lein eastwood` 🆗

### CODE REVIEW

File: test/cljam/io/util_test.clj [hunk 1]
```diff
@@ -57,24 +59,80 @@
...
     "foo.BW" :bigwig)
   (are [?dir]
-      (are [?path] (thrown? Exception (io-util/file-type (str ?dir ?path)))
-        "foo.bam.gz"
-        "foo.SAM.gz"
-        "foo.cram"
-        "foo.CRAM"
-        "foo.2bit.gz"
-        "foo.bcf.gz")
+       (are [?path] (thrown? Exception (io-util/file-type (str ?dir ?path)))
```
Comment (r6eve) [inline test/cljam/io/util_test.clj]: I think there's a path bug. When `?dir` is bound to `"/home/bar"` and `?path` is `"foo.bam.gz"`, `(str ?dir ?path)` becomes `/home/barfoo.bam.gz`. It's not an intentional result. Replacing `str` with `cio/file` might be the fix.

File: src/cljam/io/util.clj [hunk 2]
```diff
@@ -156,9 +157,53 @@
...
+        #"^##gff-version 3" :gff
+        #"^@HD\t" :sam
+        #"^@SQ\t" :sam
+        #"^@RG\t" :sam
+        #"^@PG\t" :sam
+        #"^@CO\t" :sam
+        #"^>\S" :fasta
+        #"^@\S" :fastq
+        #"(variable|fixed)Step" :wig
+        #"^\S+\t\d+\t\d+\t\d+\t\d+\n" :fai
```
Comment (r6eve) [inline src/cljam/io/util.clj]: Nitpick: I think the pattern could be concise a little: `#"^\S+(?:\t\d+){4}\n" :fai`

File: test/cljam/io/util_test.clj (hunk 1 above)
Comment (alumi) [reply test/cljam/io/util_test.clj]: Good catch! Fixed in 112988f

File: src/cljam/io/util.clj (hunk 2 above)
Comment (alumi) [reply src/cljam/io/util.clj]: Fixed in d2165e4

File: src/cljam/io/util.clj [hunk 3]
```diff
@@ -156,9 +157,53 @@
...
+        #"^##fileformat=VCF" :vcf
+        #"^##gff-version 3" :gff
+        #"^@HD\t" :sam
+        #"^@SQ\t" :sam
+        #"^@RG\t" :sam
+        #"^@PG\t" :sam
+        #"^@CO\t" :sam
+        #"^>\S" :fasta
+        #"^@\S" :fastq
+        #"(variable|fixed)Step" :wig
```
Comment (athos) [inline src/cljam/io/util.clj]: Just to confirm, is it intentional the leading `^` is omitted here?

File: src/cljam/io/util.clj [hunk 4]
```diff
@@ -156,9 +157,53 @@
...
+  "Tries to detect a file format based on contents of the byte array `ba`.
+  The input byte array must be larger than 4 bytes. Note that detection of some
+  formats `#{:fasta :fastq :wig :fai :bed}` is based on naive heuristics and
+  thus can fail."
+  [^bytes ba]
+  (let [s (String. ba 0 (Math/min (int 64) (alength ba)))
+        i (-> (ByteBuffer/wrap ba)
+              (.order ByteOrder/LITTLE_ENDIAN)
+              (.getInt)
+              (bit-and 0xffffffff))]
```
Comment (athos) [inline src/cljam/io/util.clj]: I didn't know this trick to coerce signed int to unsigned long 😲
But `Integer/toUnsignedLong` is a little more readable for this very purpose...?

File: src/cljam/io/util.clj (hunk 3 above)
Comment (alumi) [reply src/cljam/io/util.clj]: Yes, that's intentional. 

Wiggle files can be started with one of the followings:
- (optional) comment `#...`
- (optional) browser directive `browser ...`
- (optional) track definition `track ...`
- step declaration `variableStep ...`, `fixedStep ...`

First three are not specific to the format. Comments `#...` can be seen in VCF, `browser` and `track` are in BED.

[It seems like](http://genome.ucsc.edu/goldenPath/help/wiggle) wiggle format allows leading whitespaces for data lines. I'm not sure that leading whitespaces of step declaration lines are also permitted but our current implementation [supports](https://github.com/chrovis/cljam/blob/3c0371e9892f5541ebc91a61168bf29e7ecfba94/src/cljam/io/wig.clj#L83) them.

File: src/cljam/io/util.clj (hunk 4 above)
Comment (alumi) [reply src/cljam/io/util.clj]: Thanks! Fixed in 2c2123f.

File: src/cljam/io/util.clj (hunk 3 above)
Comment (athos) [reply src/cljam/io/util.clj]: > I'm not sure that leading whitespaces of step declaration lines are also permitted but our current implementation supports them.

Makes sense. OK, then 👍

**Comment by r6eve** [comment]
LGTM. Thank you for the feature.

**Comment by alumi** [comment]
Thank you for reviewing @athos @r6eve! 😺

**Pull request merged**